### ISSUE TITLE
discuss how best to get resources from squad using a command

### ISSUE THREAD

**Issue by jscook2345** [Linaro/squad-client#72]
We're trying to push back some functionality from scripts to `squad-client`. I'm trying to determine the best way to do this. This mainly revolves around getting the latest builds and their data.

In all three options we would want the ability to:
- Get the data for the latest X builds
- Get the data for a list of X builds

Three options:

## 1 Create a new program (`squad-list`) that is only responsible for dumping out data from `squad`

example usage:

```
env SQUAD_URL=? SQUAD_TOKEN=? squad-list builds --group=lkft --project=linux-next-master --count=2 --finished [BUILD, ... BUILDX]
```

This would output descriptive / declarative json similar to the following:
```
{
    { "group" : "lkft", "project" : "linux-next-master", "build" : "20201008" },
    { "group" : "lkft", "project" : "linux-next-master", "build" : "20201004" },
}
```

We could than query back to `squad` fetching the data we want. We would need to implement similar calls for all other `squad` resources.

## 2 Do the same as above, but use `sub-parser` and hope we do not need to go deeper than `two`.

example usage:

```
env SQUAD_URL=? SQUAD_TOKEN=? squad-client builds list --group=lkft --project=linux-next-master --count=2 --finished [BUILD, ... BUILDX]
```

The problem here, is we're going several commands deep and python's `argparse` does not handle that well (even sub-commands are a little hard to follow with `argparse`). We'd likely need to do something like:

```
... squad-client list-builds / build-list ...
````

instead to fix that.

We could than query back to `squad` fetching the data we want. We would need to implement similar calls for all other `squad` resources.

## 3 Do a batch dump of data and let the user select what they want.

Instead of fetching and transforming small amounts of data at a time, we could do a big data dump of results for a build and let the user than analyze that for their report.

example usage:
```
env SQUAD_URL=? SQUAD_TOKEN=? squad-client build --group=lkft --project=linux-next-master [BUILD, ... BUILDX]
```

This would output descriptive / delcarative json or csv similar to the following:
```
❯ head -n 10 dataset.csv
group.name,project.name,build.name,env.name,suite.name,test.name
lkft,linux-next-master,next-20200408,juno-r2-compat,ltp-syscalls-tests,ltp-syscalls-tests/openat02
lkft,linux-next-master,next-20200408,juno-r2-compat,ltp-syscalls-tests,ltp-syscalls-tests/chown05
lkft,linux-next-master,next-20200408,juno-r2-compat,ltp-syscalls-tests,ltp-syscalls-tests/fanotify10
lkft,linux-next-master,next-20200408,juno-r2-compat,ltp-syscalls-tests,ltp-syscalls-tests/dup03
lkft,linux-next-master,next-20200408,juno-r2-compat,ltp-syscalls-tests,ltp-syscalls-tests/adjtimex01
lkft,linux-next-master,next-20200408,juno-r2-compat,ltp-syscalls-tests,ltp-syscalls-tests/gettimeofday02
lkft,linux-next-master,next-20200408,juno-r2-compat,ltp-syscalls-tests,ltp-syscalls-tests/modify_ldt01
lkft,linux-next-master,next-20200408,juno-r2-compat,ltp-syscalls-tests,ltp-syscalls-tests/setfsgid03
lkft,linux-next-master,next-20200408,juno-r2-compat,ltp-syscalls-tests,ltp-syscalls-tests/chown01
```

From here we could use `pandas` or something similar to query the data for the report and avoid doing further api calls to squad. The hard part would be choosing what data to fetch. Everything? optional meta? optional results?

**Comment by jscook2345** [comment]
cc: @mrchapp @danrue @chaws

**Comment by chaws** [comment]
Let me work something up and i'll show you tomorrow, I'm thinking something like `squad-client list <endpoint> [--filter1 value1 [--filter2 value2]]`. 

Where `<endpoint>` can be one of "builds, projects, testjobs", etc, and `filter1` are filters used in the api, e.g. `squad-client list builds --project lkft/linux-next-master --version next-20201013`

What do you think @jscook2345 ?

**Comment by jscook2345** [comment]
@chaws 

That sounds good to me.

**Comment by jscook2345** [comment]
A lot has changed since I posted this and honestly a lot of it is confusing even to me having re-read it. I chatted with @amrohassaan in a google hang out and we decided on a smaller step first. Please see:

https://github.com/Linaro/squad-client/issues/90

Thanks!

**Comment by amrohassaan** [comment]
@chaws, I guess what Justin was asking for is enabling a filtering criteria from the command line rather than filtering squad objects to return specific fields for queried resource. For example:
curl  --header "Auth-Token: $SQUAD_TOKEN" \   
https://qa-reports.linaro.org/api/builds/xxxx/status/?finished=true&notified=true
OR 
https://qa-reports.linaro.org/api/builds/?created_at__lt=2020-11-20T18:07:48.577500Z
I guess two things will be needed: 1) document the existing squad api filter fully and 2)maybe add more filters for fields that we didn't include but are required...if any.

**Comment by amrohassaan** [comment]
After re-reading this ticket, I think @chaws proposal above is actually needed in the client as well! and we can add to each SQUAD endpoint docs the list of its filterable attrs.
